File: leetcode/merge-k-sorted-lists.py
from heapq import heappop, heappush
from typing import List, Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        dummy = ListNode(0, None)
        min_heap = []

        ptr = dummy
        length = len(lists)

        l = lists

        # construct min heap
        for i in range(length):
            if l[i]:
                heappush(min_heap, (l[i].val, i))

        while min_heap:
            _, idx = heappop(min_heap)

            ptr.next = l[idx]
            l[idx] = l[idx].next

            ptr = ptr.next
            ptr.next = None

            if l[idx]:
                heappush(min_heap, (l[idx].val, idx))

        return dummy.next

File: leetcode/merge-k-sorted-lists.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * Definition for singly-linked list.
 */
 struct ListNode {
     int val;
     ListNode *next;
     ListNode() : val(0), next(nullptr) {}
     ListNode(int x) : val(x), next(nullptr) {}
     ListNode(int x, ListNode *next) : val(x), next(next) {}
 };
 
class Solution
{
 public:
    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        auto n = lists.size();
        ListNode *dummy, *res;

        vector<pair<int, int>> heap;
        int i;

        for (i = 0; i < n; ++i) {
            if (lists[i]) {
                heap.push_back({-lists[i]->val, i});
                push_heap(heap.begin(), heap.end());
            }
        }

        dummy = new ListNode(0);
        res = dummy;

        while (!heap.empty()) {
            i = heap.front().second;

            res->next = new ListNode(-heap.front().first);
            res = res->next;

            pop_heap(heap.begin(), heap.end());
            heap.pop_back();

            lists[i] = lists[i]->next;

            if (lists[i]) {
                heap.push_back({-lists[i]->val, i});
                push_heap(heap.begin(), heap.end());
            }
        }

        res = dummy->next;
        delete dummy;

        return res;
    }
};
